src/main/scala/com/fulcrumgenomics/bam/TrimPrimers.scala

```diff
@@ -85,71 +91,117 @@ class TrimPrimers
     val outSortOrder = sortOrder.getOrElse(in.getFileHeader.getSortOrder)
     val outHeader = in.getFileHeader.clone()
     outHeader.setSortOrder(outSortOrder)
-    val out = new SAMFileWriterFactory().setCreateIndex(true).makeWriter(outHeader, outSortOrder == SortOrder.queryname, output.toFile, null)
+
+    // Setup the outputs depending on whether or not we have a reference file
+    val (sorter: Option[SortingCollection[SAMRecord]], write: (SAMRecord => Unit), out: SAMFileWriter) = ref match {
+      case Some(path) =>
+        val sorter = buildSorter(SortOrder.coordinate, outHeader)
+        val out = new SAMFileWriterFactory().setCreateIndex(true).makeWriter(outHeader, outSortOrder == SortOrder.coordinate, output.toFile, null)
+        (Some(sorter), sorter.add _, out)
+      case None =>
+        val out = new SAMFileWriterFactory().setCreateIndex(true).makeWriter(outHeader, outSortOrder == SortOrder.queryname, output.toFile, null)
+        (None, out.addAlignment _, out)
+    }
```

Comment:
There's a lot of conditional sorting going on, which now I understand reading through the commits, but perhaps a few more comments?

1. We need the reads to be queryname sorted prior to trimming.
2. If the reference is given, then we need to sort by coordinate after trimming to reset MD/NM.
3. If the reference is given, and the output sort order is coordinate (either specified via clp, or unspecified and in the input), then the records will be "presorted" when adding to the output after resetting MD/NM, otherwise we will need the output to sort.
4. If no reference is given, then the output is "presorted" if it is queryname.

So complex!